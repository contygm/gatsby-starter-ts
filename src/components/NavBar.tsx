import * as React from 'react';
import { useState } from 'react';
import { Link } from 'gatsby';
import { StaticImage } from 'gatsby-plugin-image';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMagnifyingGlass } from '@fortawesome/free-solid-svg-icons';

/**
 * @description Site wide navigation used in layout. This nobile friendly
 * navigation has the logo on the right, navigation links center, and a
 * search bar on the right.
 */
const NavBar = () => {
    const [showBurger, setShowBurger] = useState('');

    /**
     * @description Set active class for 'hamburger' menu. When active,
     * the mobile navigation options will show
     */
    function handleUpdateShowBurger() {
        const newVal = showBurger === '' ? 'is-active' : '';
        return setShowBurger(newVal);
    }

    return (
        <nav
            className="navbar is-primary"
            role="navigation"
            aria-label="main navigation"
        >
            {/* Logo Image */}
            <div className="navbar-brand">
                <Link
                    className="navbar-item"
                    to="/"
                >
                    <StaticImage
                        src="../images/bulma-logo-white.png"
                        alt="Logo"
                        placeholder="blurred"
                        className="nav-side-width"
                        layout="constrained"
                    />
                </Link>
                {/* Mobile Hamburger Graphic - following Bulma docs */}
                <button
                    className={`navbar-burger ${showBurger}`}
                    aria-label="menu"
                    aria-expanded="false"
                    onClick={handleUpdateShowBurger}
                    data-testid="siteNav"
                >
                    <span aria-hidden="true"></span>
                    <span aria-hidden="true"></span>
                    <span aria-hidden="true"></span>
                </button>
            </div>
            {/* Navigation links to main pages */}
            <div
                id="siteNav"
                className={`navbar-menu ${showBurger}`}
            >
                <div className="navbar-start">
                    <Link
                        to="/"
                        className="navbar-item"
                    >
                        Home
                    </Link>
                    <Link
                        to="/"
                        className="navbar-item"
                    >
                        Blog
                    </Link>
                    <Link
                        to="/"
                        className="navbar-item"
                    >
                        Wiki
                    </Link>
                    <Link
                        to="/404"
                        className="navbar-item"
                    >
                        Glossary
                    </Link>
                </div>
                {/* Search bar */}
                <div className="navbar-end">
                    <div className="field nav-side-width">
                        <p className="control has-icons-left has-icons-right">
                            <input
                                className="input is-rounded is-small"
                                type="text"
                                placeholder="Search..."
                            />

                            <span className="icon is-small is-left">
                                <FontAwesomeIcon icon={faMagnifyingGlass} />
                            </span>
                        </p>
                    </div>
                </div>
            </div>
        </nav>
    );
};

export default NavBar;
