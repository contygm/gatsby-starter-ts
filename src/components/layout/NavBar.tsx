import React, { useState } from 'react';
import { Link } from 'gatsby';
import { StaticImage } from 'gatsby-plugin-image';

/**
 * Site wide navigation used in layout. This nobile friendly
 * navigation has the logo on the right, navigation links center, and a
 * search bar on the right.
 * @component
 */
export const NavBar = () => {
    const [showBurger, setShowBurger] = useState('');

    /**
     * Toggle active class for 'hamburger' menu. When active,
     * the mobile navigation will show
     * @function handleUpdateShowBurger
     * @memberof NavBar
     */
    function handleUpdateShowBurger() {
        const newVal = showBurger === '' ? 'is-active' : '';
        return setShowBurger(newVal);
    }

    return (
        <nav
            className="navbar is-primary"
            role="navigation"
            aria-label="main navigation"
        >
            {/* Logo Image */}
            <div className="navbar-brand">
                <Link
                    className="navbar-item"
                    to="/"
                >
                    <StaticImage
                        src="../../images/bulma-logo-white.png"
                        alt="Logo"
                        placeholder="blurred"
                        className="nav-side-width"
                        layout="constrained"
                        data-cy="nav-link-logo"
                    />
                </Link>
                {/* Mobile Hamburger Graphic - following Bulma docs */}
                <button
                    className={`navbar-burger ${showBurger}`}
                    aria-label="menu"
                    aria-expanded="false"
                    onClick={handleUpdateShowBurger}
                    data-testid="siteNav"
                    data-cy="nav-hamburger"
                >
                    <span aria-hidden="true"></span>
                    <span aria-hidden="true"></span>
                    <span aria-hidden="true"></span>
                </button>
            </div>
            {/* Navigation links to main pages */}
            <div
                id="siteNav"
                className={`navbar-menu ${showBurger}`}
            >
                <div className="navbar-start">
                    <Link
                        to="/"
                        className="navbar-item"
                        data-cy="nav-link-home"
                    >
                        Home
                    </Link>
                    <Link
                        to="/blog"
                        className="navbar-item"
                        data-cy="nav-link-blog"
                    >
                        Blog
                    </Link>
                    <Link
                        to="/wiki"
                        className="navbar-item"
                        data-cy="nav-link-wiki"
                    >
                        Wiki
                    </Link>
                    <Link
                        to="/glossary"
                        className="navbar-item"
                        data-cy="nav-link-glossary"
                    >
                        Glossary
                    </Link>
                </div>
            </div>
        </nav>
    );
};
