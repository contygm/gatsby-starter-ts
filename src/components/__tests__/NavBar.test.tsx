import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { NavBar } from '../NavBar';

describe('NavBar', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<NavBar />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('hamburger button changes classname', () => {
        const { getByTestId } = render(<NavBar />);

        expect(getByTestId('siteNav').classList.contains('is-active')).toBe(
            false
        );
        fireEvent.click(getByTestId('siteNav'));
        expect(getByTestId('siteNav').classList.contains('is-active')).toBe(
            true
        );
        fireEvent.click(getByTestId('siteNav'));
        expect(getByTestId('siteNav').classList.contains('is-active')).toBe(
            false
        );
    });
});
