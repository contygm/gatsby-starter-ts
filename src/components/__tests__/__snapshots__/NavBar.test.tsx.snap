// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders correctly 1`] = `
<DocumentFragment>
  <nav
    aria-label="main navigation"
    class="navbar is-primary"
    role="navigation"
  >
    <div
      class="navbar-brand"
    >
      <a
        class="navbar-item"
        href="/"
      >
        <img
          alt="Logo"
          data-cy="nav-link-logo"
          layout="constrained"
          placeholder="blurred"
          src="../images/bulma-logo-white.png"
        />
      </a>
      <button
        aria-expanded="false"
        aria-label="menu"
        class="navbar-burger "
        data-cy="nav-hamburger"
        data-testid="siteNav"
      >
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
        <span
          aria-hidden="true"
        />
      </button>
    </div>
    <div
      class="navbar-menu "
      id="siteNav"
    >
      <div
        class="navbar-start"
      >
        <a
          class="navbar-item"
          data-cy="nav-link-home"
          href="/404"
        >
          Home
        </a>
        <a
          class="navbar-item"
          data-cy="nav-link-blog"
          href="/blog"
        >
          Blog
        </a>
        <a
          class="navbar-item"
          data-cy="nav-link-wiki"
          href="/wiki"
        >
          Wiki
        </a>
        <a
          class="navbar-item"
          data-cy="nav-link-glossary"
          href="/glossary"
        >
          Glossary
        </a>
      </div>
      <div
        class="navbar-end"
      >
        <div
          class="field nav-side-width"
        >
          <p
            class="control has-icons-left has-icons-right"
          >
            <input
              class="input is-rounded is-small"
              placeholder="Search..."
              type="text"
            />
            <span
              class="icon is-small is-left"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-magnifying-glass "
                data-icon="magnifying-glass"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352c79.5 0 144-64.5 144-144s-64.5-144-144-144S64 128.5 64 208s64.5 144 144 144z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </p>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
